src/v/cloud_storage/topic_manifest.cc

```diff
@@ -39,8 +39,909 @@
+            return handle_try_read(key, val);
+        }
+
+        auto try_read(std::string_view key, bool val) const -> bool override {
+            return handle_try_read(key, val);
+        }
+        auto try_read(std::string_view key, std::string_view val) const
+          -> bool override {
+            return handle_try_read(key, val);
+        }
+        auto try_read(std::string_view key, int64_t val) const
+          -> bool override {
+            return handle_try_read(key, val);
+        }
+        auto try_read(std::string_view key, uint64_t val) const
+          -> bool override {
+            return handle_try_read(key, val);
+        }
+
+        explicit impl(Klass& obj) noexcept
+          : obj_{&obj} {}
+
+    private:
+        // this points to the object to populate
+        Klass* obj_;
+    };
+
+public:
+    /// create a mapping for an Klass from a collection of
+    /// type_mapping/aggregate_type_mapping. \param Klass is used to enable
+    /// constructor template argument deduction, only the type is meaningfull.
+    /// pass a nullptr like this: static_cast<T const*>(nullptr)
+    consteval explicit root_type_mapping(Klass const*, Fn... fields)
+      : fields_{fields...} {}
+
+    /// prepare obj for deserialization, and create a new context that will
+    /// handle key:value
+    auto start(Klass& obj) const
+      -> std::unique_ptr<const object_mapping_context> {
+        // normalize the fields of obj for deserialization, return a context
+        // pointing at this object
+        std::apply(
+          [&](auto const&... field) {
+              (field.prepare_for_deserialization(obj), ...);
+          },
+          fields_);
+        return std::make_unique<const impl>(obj);
+    }
+
+    /// serialize all the fields of obj, it is expected that the caller will
+    /// wrap this in StartObject EndObject
+    void write_fields(Klass const& obj, writer_t& w) const {
+        std::apply(
+          [&](auto const&... field) { (field.write(obj, w), ...); }, fields_);
+    }
+
+private:
+    std::tuple<Fn...> fields_;
+};
+
+using rtp = cluster::remote_topic_properties;
+/// table of mappings for cluster::remote_topic_properties
+template<>
+struct field_mappings_for_t<cluster::remote_topic_properties> {
+    constexpr static auto value = root_type_mapping{
+      static_cast<cluster::remote_topic_properties const*>(nullptr),
+      type_mapping{
+        "remote_revision",
+        &rtp::remote_revision,
+        [](model::initial_revision_id id) { return int64_t(id); },
+        [](int64_t id) { return model::initial_revision_id(id); }},
+      type_mapping{
+        "remote_partition_count",
+        &rtp::remote_partition_count,
+        [](int32_t in) { return in; },
+        [](int32_t in) { return in; }},
+    };
+};
+
+using tp = cluster::topic_properties;
+/// table of the json mappings for each field of cluster::topic_properties
+template<>
+struct field_mappings_for_t<cluster::topic_properties> {
```

Comment:
I haven't reviewed this properly but I'll add the following to the list of concerns:
* Field mappings don't seem to be able to handle any type of versioning. It is not hard to imagine a case where in the future we might want to have version dependent deserializing.
* What if field type changes between versions? E.g. was an int, decided to use a string in a future version?

Reply:
- `Can/is this generic enough to be pulled out into its own reusable json serialization class`
  - kind of. It lacks support for sequences, but it seems feasible to describe them. the real problem is tristate; their current serialization breaks composability. if you don't see a key, was it because the tristate is disabled or because the property did not exist yet? right now, I'm serializing (unpublished commit) an array with the disabled tristate, but if we want to add support for sequences of tristates we need to change the encoding to an object like `{"state" : "disabled"}` `{"state": "empty"}` `{"state": "value", "value" = ...}`. kinda ugly but with a version change we could do it 
-  `Wondering if you'd considered just adding some to_manifest_json`
  -  wdym? can't find to_manifest_json in the codebase
- `Field mappings don't seem to be able to handle any type of versioning`
  - i can add this to `root_type_mapping`, def a useful addition
- `What if field type changes between versions? E.g. was an int, decided to use a string in a future version?`
  - in the hypothesis that we do incompatible type changes, we have two cases:
     1. old redpanda reading data from new repanda: std::runtime_error, failing the future. we don't have a policy for making data readable from old versions. I can add a feature check to ensure that the cluster is upgraded before switching to a new format
     2. new redpanda reading old data: some support. the from_json functor can handle string/bool/int. a T can be read as optional<T>, and they can be read as tristate<T>. a tristate<T> can be read as an optional<T>, but they can't be read as T (unless you modify type_mapping to not signal error on `null` values). moving to an aggregate means adapting aggregate_type_mapping to handle this.
     
when writing this, I had these goals:
1. keep a json format that can be modified without tooling
2. build upon the pre-existing encoding
3. have some sort of serde_fields() affinity so that it's possible to ensure that the mappings are up to date
3.b A new property should mean just adding a type_mapping or an aggregate_type_mapping/root_object_mapping pair
4. don't rewrite the whole thing, specialize it for this usecase

initially, I just had a tuple of type_mappings, which was relatively simple. but once reached remote_topic_properties I had to increase the complexity to encode it, or not encode it but move the problem into the future and add some context checks like "the number of mappings should be 1-less the number of fields". I realize that 3 is in a bit of contraposition with the other goals, and 4 is not exactly achievable with this code.

we could move to binary encoding, but maybe there is value in human readable, and I would be almost ok with keeping the chain of manual key-value writes if we had a mechanism to check for completeness, like a test that automatically adds new topic properties or some sort of compile-time counter .